corehq/apps/api/tests/test_auth.py

```diff
@@ -131,6 +131,17 @@ def reactivate_key():
         self.api_key.refresh_from_db()
         self.assertIsNotNone(self.api_key.last_used)
 
+    def test_login_with_deactivated_user(self):
+        def reactivate_user():
+            self.user.is_active = True
+            self.user.save()
+        self.user.is_active = False
+        self.user.save()
+        self.addCleanup(reactivate_user)
```

Comment:
Hey, just wanted to leave a comment that this approach is risky. I know you're following what the other tests are doing here, but we generally want to be avoiding tests with side-effects that could change the behavior of other tests. If we had forgotten to include the `reactivate_key` call (or if that call fails for some reason), this test may still pass. The really bad scenario is when all tests also pass, but only because this test randomly happened to be the last test called in this test suite. It opens the possibility that, at a later date, it gets executed first, rather than last, and it ends up being fairly confusing to understand the failure.

I don't think it's necessary to change anything now, but I'd recommend if we do create a test like this in the future, that we create fresh users for each test, rather than recycling the same user between tests. That allows us to forget things like cleaning up, because the user object won't affect anything else.

Reply:
Thank you Matt! Are you suggesting, if we need to alter attributes of on object, then we always create a fresh object for each test, rather then reuse the same object?

Reply:
Correct. In this case, that would be moving the user creation from `setupClass` to `setup`. I'd always start with creating objects in `setup`, and only relying on `setupClass` if performance proves to be an issue. For objects that won't get modified, it's fine to rely on `setupClass`